### PR TITLE
Add suppressApplicationTitle as boolean

### LINKED ISSUE #2645: "suppressApplicationTitle" doesn't work (or exist.)
# Environment

```
Windows 10.0.18362.0
Terminal: 0.4.2382.0
```

# Steps to reproduce

1. Add custom, permanent tab titles for some of the profiles:
```
  "tabTitle" : "gateway",
  "suppressApplicationTitle": true,
```
2. Start a new application in these tabs.

I took the syntax from the PR by @zadjii-msft: https://github.com/microsoft/terminal/pull/2373#issuecomment-520832247

# Expected behavior

Tab titles don't change, they keep the value specified in `tabTitle`, just like they used to in Terminal 0.3.x.

# Actual behavior

Tab titles change to reflect the currently running program.

Here I'm running three different applications using `npm start`. I've set up a different, useful `tabTitle` for all of these tabs, but all I see is this:

image

# Note

From #2373: _"d. We'll add suppressApplicationTitle when someone complains"_ <- Here I am. :)

### PR DESCRIPTION
**Pull request by cinnamon-msft**
## Summary of the Pull Request
Add suppressApplicationTitle as a setting

 
## PR Checklist
* [x] Closes #2645
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Validation Steps Performed
If there's no `tabTitle`, `suppressApplicationTitle` does nothing
If `suppressApplicationTitle` is set to false, `tabTitle` acts as normal
If they both exist, `tabTitle` persists as the tab title

### CODE REVIEW

File: doc/cascadia/SettingsSchema.md [hunk 1]
```diff
@@ -45,6 +45,7 @@ Properties listed below are specific to each unique profile.
 | `foreground` | Optional | String | | Sets the foreground color of the profile. Overrides `foreground` set in color scheme if `colorscheme` is set. Uses hex color format: `"#rrggbb"`. |
 | `icon` | Optional | String | | Image file location of the icon used in the profile. Displays within the tab and the dropdown menu. See [Background Images and Icons](./SettingsSchema.md#background-images-and-icons) below for help on specifying your own icons |
 | `scrollbarState` | Optional | String | | Defines the visibility of the scrollbar. Possible values: `"visible"`, `"hidden"` |
+| `suppressApplicationTitle` | Optional | Boolean | | When set to `true`, `tabTitle` overrides the default title of the tab. When set to `false`, `tabTitle` behaves as normal. |
```
Comment (miniksa) [inline doc/cascadia/SettingsSchema.md]: Can you explain a bit further that it's suppressing any title change messages from the client application? It's took me substantially longer to realize this than it probably should have.

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 2]
```diff
@@ -563,7 +563,19 @@ namespace winrt::TerminalApp::implementation
...
-        tab->SetTabText(newTabTitle);
+        const auto lastFocusedProfile = tab->GetFocusedProfile().value();
+
+        auto matchingProfile = _settings->FindProfile(lastFocusedProfile);
+        if (matchingProfile)
+        {
+            if (matchingProfile->GetSuppressApplicationTitle())
+            {
+                auto profileTabTitle = matchingProfile->GetTabTitle();
+                newTabTitle = profileTabTitle.empty() ? matchingProfile->GetName() : profileTabTitle;
```
Comment (miniksa) [inline src/cascadia/TerminalApp/TerminalPage.cpp]: Why does this operate on this side of the equation instead of preventing tab->SetFocusedTitle() from storing a client application set title in the first place?

File: src/cascadia/TerminalApp/Profile.cpp [hunk 3]
```diff
@@ -806,6 +813,21 @@ std::wstring_view Profile::GetName() const noexcept
     return _name;
 }
 
+// Method Description:
+// - Returns the tab title, if one is set. Otherwise returns the empty string.
+// Return Value:
+// - this profile's tab title, if one is set. Otherwise returns the empty string.
+std::wstring_view Profile::GetTabTitle() const noexcept
+{
+    return _tabTitle.has_value() ? std::wstring_view{ _tabTitle.value().c_str(), _tabTitle.value().size() } :
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/Profile.cpp]: What about when the user is using `name` as the tab title? Should just `name` +`suppressApplicationTitle:true` result in the profile's name suppressing the title?

File: src/cascadia/TerminalSettings/terminalsettings.h [hunk 4]
```diff
@@ -88,6 +88,9 @@ namespace winrt::Microsoft::Terminal::Settings::implementation
         hstring StartingTitle();
         void StartingTitle(hstring const& value);
 
+        bool SuppressApplicationTitle();
+        void SuppressApplicationTitle(bool value);
```
Comment (zadjii-msft) [inline src/cascadia/TerminalSettings/terminalsettings.h]: Do these need to be here at all? Doesn't look like they're being used by `TerminalPage` at all.

Typically this settings class is only used for `ICoreSettings` or `IControlSettings`, and I don't think `SuppressApplicationTitle` is either of those, it's more of an _app_ setting

**Comment by thongtrinh2204** [comment]
hi @cinnamon-msft  @zadjii-msft do we have any update on it? i really need this feature for WSL.

File: src/cascadia/TerminalCore/TerminalApi.cpp [hunk 5]
```diff
@@ -365,11 +365,23 @@ bool Terminal::EraseInDisplay(const DispatchTypes::EraseType eraseType)
...
+    // Set the title on Terminal load
+    if (_title.empty())
+    {
+        _title = title;
+        _pfnTitleChanged(title);
+    }
+
     _title = title;
 
+    if (_suppressApplicationTitle)
```
Comment (zadjii-msft) [inline src/cascadia/TerminalCore/TerminalApi.cpp]: Couldn't we just _not_ call `_pfnTitleChanged` when `_suppressApplicationTitle` is true? That would negate the need to the `Terminal` core to know about startingTitle, it would just _not_ send title change events at all.

Maybe the `_pfnTitleChanged` needs to be called on the _first_ time through this method, but not on subsequent calls...

File: src/cascadia/TerminalCore/Terminal.hpp [hunk 6]
```diff
@@ -162,12 +162,14 @@ class Microsoft::Terminal::Core::Terminal final :
     std::unique_ptr<::Microsoft::Console::VirtualTerminal::TerminalInput> _terminalInput;
 
     std::wstring _title;
+    std::wstring_view _startingTitle;
```
Comment (zadjii-msft) [inline src/cascadia/TerminalCore/Terminal.hpp]: I'm _mildly concerned_ about storing this as a `wstring_view`. Won't this explode when the method that sets this goes out of scope, as the backing `wstring` gets freed?

File: src/cascadia/TerminalCore/TerminalApi.cpp (hunk 5 above)
Comment (cinnamon-msft) [reply src/cascadia/TerminalCore/TerminalApi.cpp]: If we don't call `_pfnTitleChanged` when `_suppressApplicationTitle` is true, refocusing on the tab/pane within the tab (`term.GotFocus()` in TerminalPage.cpp) will call that tab's stored title, which is stored as the application title.

File: src/cascadia/TerminalApp/Profile.cpp [hunk 7]
```diff
@@ -93,6 +94,7 @@ Profile::Profile(const std::optional<GUID>& guid) :
     _selectionBackground{},
     _colorTable{},
     _tabTitle{},
+    _suppressApplicationTitle{},
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/Profile.cpp]:
```suggestion
    _suppressApplicationTitle{ false },
```

File: src/cascadia/UnitTests_TerminalCore/MockTermSettings.h [hunk 8]
```diff
@@ -50,15 +52,21 @@ namespace TerminalCoreUnitTests
...
         void CopyOnSelect(bool copyOnSelect) { _copyOnSelect = copyOnSelect; }
+        void StartingTitle(winrt::hstring const& value) { _startingTitle = value; }
+        void SuppressApplicationTitle(bool suppressApplicationTitle) { _suppressApplicationTitle = suppressApplicationTitle; }
         void SelectionBackground(uint32_t) {}
 
         // other unimplemented methods
-        void SetColorTableEntry(int32_t /* index */, uint32_t /* value */) {}
+        void SetColorTableEntry(int32_t /* index */, uint32_t /* value */)
+        {
+        }
```
Comment (carlos-zamora) [inline src/cascadia/UnitTests_TerminalCore/MockTermSettings.h]: This change seems unrelated. Did the formatter catch it? Revert it if possible.

File: src/cascadia/TerminalApp/Profile.cpp (hunk 7 above)
Comment (cinnamon-msft) [reply src/cascadia/TerminalApp/Profile.cpp]: Isn't this only if it's set false as default?

File: src/cascadia/TerminalCore/TerminalApi.cpp (hunk 5 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalCore/TerminalApi.cpp]: It seems like we could just set `_title` to `StartingTitle` as early as possible (when we apply the settings), and then we never have to make the check here on 368-373. Then, we can just simplify the code in here to:

```
if (!_suppressApplicationTitle)
{
    _title = title;
    if (_pfnTitleChanged)
    {
        _pfnTitleChanged(_title)
    }
}
return true;
```

and that's it. We never update `_title`, we never check if it's empty (settings always sets it), we don't have to store `_startingTitle` separately :smile:

File: src/cascadia/TerminalCore/TerminalApi.cpp [hunk 9]
```diff
@@ -365,13 +365,23 @@ bool Terminal::EraseInDisplay(const DispatchTypes::EraseType eraseType)
...
 {
-    _title = title;
-
-    if (_pfnTitleChanged)
+    // Set the title on Terminal load
+    if (_title.empty())
     {
+        _title = title;
         _pfnTitleChanged(title);
     }
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalCore/TerminalApi.cpp]: I still believe you can remove all this, since the _title should never be empty now that UpdateSettings is taking care of it on startup

Comment (cinnamon-msft) [reply src/cascadia/TerminalCore/TerminalApi.cpp]: Removing this causes the tab title to be blank until re-focused on when suppressApplicationTitle is set to true.

Comment (cinnamon-msft) [reply src/cascadia/TerminalCore/TerminalApi.cpp]: Hah! Just kidding, just had a slow-to-load tab :)

Comment (cinnamon-msft) [reply src/cascadia/TerminalCore/TerminalApi.cpp]: Wait, no. It happens if I haven't previously loaded that profile in another tab. If I have two of the same profiles open, it behaves fine.

Comment (DHowett-MSFT) [reply src/cascadia/TerminalCore/TerminalApi.cpp]: That shouldn't be the case ... o_O the title should never be empty because it was getting set by `updateSettings`

oh crap i just realized i made you make a bad/annoying change

Comment (DHowett-MSFT) [reply src/cascadia/TerminalCore/TerminalApi.cpp]: okay, if you roll back to commit 5badf2c you'll be fine. If you don't have any local changes you care about:

```
git reset --hard 5badf2c 
git push -f
```

**Comment by DHowett-MSFT** [comment]
My comment no longer applies, and we reverted the code that make it happen because it was actually worse
